### PR TITLE
Customsearch API swagger spec

### PR DESCRIPTION
**Pull request by q3blend**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+					"description": "A Boolean value that indicates whether the specified query has adult intent. The value is true if the query has adult intent; otherwise, false.",
+					"readOnly": true,
+					"type": "boolean"
+				},
+				"askUserForLocation": {
+					"description": "A Boolean value that indicates whether Bing requires the user's location to provide accurate results. If you specified the user's location by using the X-MSEdge-ClientIP and X-Search-Location headers, you can ignore this field. For location aware queries, such as \"today's weather\" or \"restaurants near me\" that need the user's location to provide accurate results, this field is set to true. For location aware queries that include the location (for example, \"Seattle weather\"), this field is set to false. This field is also set to false for queries that are not location aware, such as \"best sellers\".",
+					"readOnly": true,
+					"type": "boolean"
+				},
+				"isTransactional": {
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: >isTransactional [](start = 5, length = 15)

Why not include a description?

Comment (q3blend) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: isTransactional was actually an unused param. I've removed it.

Comment (mcardosos) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Are you also removing it for CustomSearch? or only from webSearch swagger?

---
In reply to: [154246597](https://github.com/Azure/azure-rest-api-specs/pull/2065#discussion_r154246597) [](ancestors = 154246597)

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+				"text": {
+					"readOnly": true,
+					"type": "string"
+				}
+			}
+		},
+		"WebMetaTag": {
+			"description": "Defines a webpage's metadata.",
+			"x-ms-discriminator-value": "Web/MetaTag",
+			"type": "object",
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Which is the discriminator property here?

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+					"$ref": "#/definitions/QueryContext",
+					"readOnly": true
+				},
+				"totalEstimatedMatches": {
+					"description": "The estimated number of webpages that are relevant to the query. Use this number along with the count and offset query parameters to page the results.",
+					"readOnly": true,
+					"type": "integer",
+					"format": "int64"
+				},
+				"isFamilyFriendly": {
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: >isFamilyFriendly [](start = 5, length = 16)

Please provide a description

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    "q": "cortana"
+  },
+  "responses": {
+    "200": {
+      "headers": {
+        
+      },
+      "body": {
+        "_type": "SearchResponse",
+        "webPages": {
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json]: >webPages [](start = 9, length = 8)

Needs its discriminator

**Review by mcardosos — CHANGES_REQUESTED** [review]
Left some comments.
From a code generation point of view, this swagger would greatly benefit from [x-ms-parameter-grouping](https://github.com/stankovski/AutoRest/blob/master/Documentation/swagger-extensions.md). This provides a better user experience and avoids a method with a gigantic signature with 15+ parameters.
Also there are [model validation errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/309812187#L694)

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json (hunk 3 above)
Comment (q3blend) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Not stated in documentation. This is an indirect reference. There is no description for it in websearch spec

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json (hunk 4 above)
Comment (q3blend) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json]: What do you mean by discriminator? What do I need to add? This is the response to customsearch API copied over.

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json (hunk 2 above)
Comment (q3blend) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: What is a discriminator property?

Comment (sudhansu88) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: It should be _type
 "_type": {
          "type": "string"
        }

Based on https://github.com/Azure/oav/pull/171, We don't need to specify discriminator property for every child model.

**Comment by sudhansu88** [comment]
Please add below to readme

## Suppressions 

Suppressing errors due to API design: 
``` yaml 
directive:
  - suppress: R3016
    reason: _type is a polymorphic discriminator that can't be changed.

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json (hunk 4 above)
Comment (sudhansu88) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json]: https://github.com/Azure/oav/pull/171

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json (hunk 3 above)
Comment (sudhansu88) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Usually we don't have description for easily understood schema fields

 THis is same as below
https://github.com/Azure/azure-rest-api-specs/blob/current/specification/cognitiveservices/data-plane/WebSearch/v1.0/WebSearch.json

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json (hunk 2 above)
Comment (mcardosos) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Is this a base type?

Comment (sudhansu88) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: This is a derived type

Comment (mcardosos) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: But it has no base property, why is the discriminator value needed?

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          "_type": "Web/WebAnswer",
+          "value": [
+            {
+              "id": "https://api.cognitive.microsoft.com/api/v7/#WebPages.0",
+              "name": "Cortana Skills - microsoft.com",
+              "url": "https://www.microsoft.com/en-us/cortana/skills/all",
+              "displayUrl": "https://www.microsoft.com/en-us/cortana/skills/all",
+              "snippet": "“Hey Cortana, ask Akinator to guess who I am thinking of.” “Hey Cortana, ask Akinator to start.” “Hey Cortana, ask Akinator to play a new round.” “Hey ...",
+              "dateLastCrawled": "2017-11-27T14:12:00.0000000Z"
+            }
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/v1.0/examples/SuccessfulQueryRequest.json]: This model also needs a discriminator (very deep inside, one of is base models requires a discriminator property)

File: specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json (hunk 2 above)
Comment (sudhansu88) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Will remove
@q3blend : Could you please remove "x-ms-discriminator-value": "Web/MetaTag" as it truly not needed now

Comment (q3blend) [reply specification/cognitiveservices/data-plane/CustomSearch/v1.0/CustomSearch.json]: Removed

File: specification/cognitiveservices/data-plane/CustomSearch/readme.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+These settings apply only when `--csharp` is specified on the command line.
+``` yaml $(csharp) 
+csharp: 
+  namespace: Microsoft.Azure.CognitiveServices.Search.CustomSearch
+  output-folder: $(csharp-sdks-folder)/CognitiveServices/dataPlane/Search/BingCustomSearch/Generated/CustomSearch
+  sync-methods: none
+```
+## Suppressions
+Suppressing errors due to API design:
+``` yaml $(csharp)
```
Comment (mcardosos) [inline specification/cognitiveservices/data-plane/CustomSearch/readme.md]: > $(csharp)  [](start = 8, length = 11)

Last detail: you dont need the `$(csharp)` part

**Review by mcardosos — APPROVED** [review]
Cool!
Thnaks for the awesome cooperation @q3blend  @sudhansu88
:D

**Pull request merged**